### PR TITLE
Extension to keyboard controls

### LINKED ISSUE #3517: Code changes re. Keyboard controls #3484
### Description

This relates to discussion Keyboard controls #3484
In working this though there are some code changes in SmoothControl.ts that you can have if you want them.
These extend the keyboard controls from orbit changes to orbit, pan, step size & preset views and also remove the null hit behavior.
The removal of null hit behavior should really be conditional on an exposed state, but I didn't know how to do that.
They are contained entirely within SmoothControls.ts. 

Keyboard controls are:
Arrows - orbit
Shift + Arrows - pan
Ctrl + Arrows - orbit at small step (ratio from unmodified set as a const, but could ideally be passed in)
Ctrl + Shift + Arrows - pan at small steps

Pressing Shift displays the same spot cursor as used on mouse-pan. Shift now behaves as a defacto 'about to pan' indicator. It also accidentally solved a user comment that they didn't really know where it was orbiting around.

Keys 1, 2, 3, 4 (and Numpad 1, 2, 3, 4) set theta to 0, 90, 180, 270 degrees at current phi and radius
Keys 7, 8, 9 (and Numpad 7, 8, 9) set phi to +45, 0 and -45 degrees at current theta and radius
This gives the ability to 'square up' a view which works in conjunction with the orbit keys.

Only experimental testing has been done. No automated test.

#### Live Demo
Video clips in  Keyboard controls #3484

### Version

- model-viewer: v1.12.0

### Browser Affected

- [ 1 ] Chrome, version: xx.x.xxxx.xx
- [ 1] Edge
- [ 1] Firefox
- [ 0] IE
- [ 0] Safari

### OS

- [ 0] Android
- [ 0] iOS
- [ 0] Linux
- [ 0] MacOS
- [ 1] Windows

### AR

- [ ] WebXR
- [ ] SceneViewer
- [ ] QuickLook

### PR DESCRIPTION
**Pull request by clochardM33**
### Reference Issue

Fixes #3517

This relates to discussion Keyboard controls https://github.com/google/model-viewer/discussions/3484
In our specific world we are involved with technical drawings, and now with model-viewer, 3D models. The techy nature of it lends itself to wanting to view models at specific views (the orthogonal ones are the obvious examples) and wanting to be able to put models in specific orientations (although most of the users I've observed don't realise that is what they are trying to do with the mouse painfully sliding back and forth over the same spot).
It doesn't seem like an obvious need when looking at a model of a show or a tangerine as they don't have the inherent precision that a CAD dwg brings to mind.
The current keyboard controls are exactly the right sort of thing, but they are tantalisingly short of functionality.

In working this though there are some code changes in SmoothControl.ts (and a quick one in the package list that is unrealted).
These extend the keyboard controls from orbit changes to orbit, pan, step size & preset views and also remove the null hit behavior.
The removal of null hit behavior should really be conditional on an exposed state, but I didn't know how to do that.
They are contained entirely within SmoothControls.ts.

Keyboard controls are:
Arrows - orbit
Shift + Arrows - pan
Ctrl + Arrows - orbit at small step (ratio from unmodified set as a const, but could ideally be passed in)
Ctrl + Shift + Arrows - pan at small steps

Pressing Shift displays the same spot cursor as used on mouse-pan. Shift now behaves as a defacto 'about to pan' indicator. It also accidentally solved a user comment that they didn't really know where it was orbiting around.

Keys 1, 2, 3, 4 (and Numpad 1, 2, 3, 4) set theta to 0, 90, 180, 270 degrees at current phi and radius
Keys 7, 8, 9 (and Numpad 7, 8, 9) set phi to +45, 0 and -45 degrees at current theta and radius
This gives the ability to 'square up' a view which works in conjunction with the orbit keys.

The ration of small steps to big step should probably be configurable, but as above, that is something that I can't see how to get from the top level interface down to this script. 

Only experimental testing has been done. No automated test.

Live Demo
Video clips in Keyboard controls https://github.com/google/model-viewer/discussions/3484

### CODE REVIEW

File: packages/model-viewer/src/three-components/SmoothControls.ts [hunk 1]
```diff
@@ -641,11 +660,17 @@ export class SmoothControls extends EventDispatcher {
...
-      const {cameraTarget} = scene.element;
-      scene.element.cameraTarget = '';
-      scene.element.cameraTarget = cameraTarget;
-      // Zoom all the way out.
-      this.userAdjustOrbit(0, 0, 1);
+      // 'This should probably be optional, but I don't know how to extend it to pass something into here
+
+      // No, don't move all the way out.
+      // If clicked on nothing then do nothing.
+      // Resetting on hit == null means you forfeit the position you were just in. If that position is intentional then you lose that intention.
```
Comment (elalish) [inline packages/model-viewer/src/three-components/SmoothControls.ts:667]: Can you describe the situations where you find this to be a problem? This behavior was put in to allow the user to return to the initial pan state, which is basically impossible when panning back and forth.

File: packages/model-viewer/src/three-components/SmoothControls.ts [hunk 2]
```diff
@@ -839,33 +864,168 @@ export class SmoothControls extends EventDispatcher {
...
+    let swizzkey = true;
+    switch (event.keyCode) {
+      case KeyCode.PAGE_UP:
+        // Do nothing for this
+        // this.userAdjustOrbit(0, 0, ZoomInc);
+        break;
+      case KeyCode.PAGE_DOWN:
+        // Do nothing for this
+        // this.userAdjustOrbit(0, 0, -1 * ZoomInc);
+        break;
```
Comment (elalish) [inline packages/model-viewer/src/three-components/SmoothControls.ts:997]: Can you just remove these cases if they do nothing?

File: packages/model-viewer/src/three-components/SmoothControls.ts [hunk 3]
```diff
@@ -839,33 +864,168 @@ export class SmoothControls extends EventDispatcher {
...
+      case KeyCode.NP_KEY_8:
+          this.keyPressPhiSet(Math.PI /2);
+        break;
+      case KeyCode.KEY_9:
+      case KeyCode.NP_KEY_9:
+          this.keyPressPhiSet(Math.PI / 4);
+        break;   
       default:
-        relevantKey = false;
+        swizzkey = false;
```
Comment (elalish) [inline packages/model-viewer/src/three-components/SmoothControls.ts:979]: What does `swizzkey` represent?

File: packages/model-viewer/src/three-components/SmoothControls.ts [hunk 4]
```diff
@@ -839,33 +864,168 @@ export class SmoothControls extends EventDispatcher {
...
+  }
+
+  // Adjust only the phi of the orbit
+  // (Effectively spinning around the target vertically)
+  private keyPressPhiSet(newPhi: number) {
+    const {theta, radius} = this.goalSpherical;
+    this.setOrbit(theta, newPhi, radius);
+  }
 }
+
```
Comment (elalish) [inline packages/model-viewer/src/three-components/SmoothControls.ts:1031]: Can you find a way to remove all those extra empty diff files? Not sure where those are coming from.

**Comment by clochardM33** [comment]
The review comments are great. Off the back of them I've rethought everything.

File: packages/model-viewer/src/features/controls.ts [hunk 5]
```diff
@@ -361,6 +366,18 @@ export const ControlsMixin = <T extends Constructor<ModelViewerElementBase>>(
...
+    panKeyIncrement: number = DEFAULT_PANKEYINCREMENT;
+
+    @property({type: Number, attribute: 'pan-key-multiplier'})
+    panKeyMultiplier: number = DEFAULT_PANKEYMULTIPLIER;
+
+    @property({type: Number, attribute: 'pan-mouse-multiplier'})
+    panMouseMultiplier: number = DEFAULT_PANMOUSEMULTIPLIER;
+
+    @property({type: Number, attribute: 'orbit-key-multipler'})
+    orbitKeyMultipler: number = DEFAULT_ORBITKEYMULTIPLIER;
```
Comment (elalish) [inline packages/model-viewer/src/features/controls.ts:379]: I am not excited about expanding our API surface this much for such a small feature. Let's talk through design first.

File: packages/model-viewer/src/three-components/SmoothControls.ts [hunk 6]
```diff
@@ -764,9 +783,19 @@ export class SmoothControls extends EventDispatcher {
       element.removeEventListener('pointermove', this.onPointerMove);
       element.removeEventListener('pointerup', this.onPointerUp);
       element.removeEventListener('touchmove', this.disableScroll);
-      if (this.enablePan) {
-        this.recenter(event);
+
+      /** Something for double click detection.
+       * Only fire the recenter event on double click. This prevents errant (single) click events (such as giving back focus to model-viewer) from changing the current view.
+       * If the time difference between now and the previous mouse click (not the current one, the one before that) is less than TAP_MS then call it a double click.
+      */
```
Comment (elalish) [inline packages/model-viewer/src/three-components/SmoothControls.ts:790]: This is interesting, but appears to be unrelated to the rest of this PR. Perhaps separate it? This is already a bit large.

**Review by elalish — CHANGES_REQUESTED** [review]
Thanks, but can you please respond to my comments? Even several that are marked "outdated" still apply to your current PR. I've marked each that I'd really like a response to with the "eyes" emoji.

**Comment by clochardM33** [comment]
Just want to double check that I explicitly press Re-request review when I've got to the end of comments (rather than just leaving it)? Thanks.

**Comment by elalish** [comment]
> Just want to double check that I explicitly press Re-request review when I've got to the end of comments (rather than just leaving it)? Thanks.

That would be great, yes. Most people don't bother; I get pinged for commits and comments, so those also tend to get my attention.

**Review by elalish — CHANGES_REQUESTED** [review]
You're not waiting for me here, are you? I would like to merge this, though I have a few significant concerns to address first. Go ahead and reply to the comments - once we've discussed it should be clear what changes to make.

**Comment by clochardM33** [comment]
As it happens I am waiting on your reply as I'm not sure what to do next and where we got up to.
I think I'm in the running for the Mr T blathering fool most ignorant Git user of the project.

**Comment by elalish** [comment]
Okay, no problem. Please go up to each thread I started in your code and reply to those. Even if the reply is "I'm not sure what you want", that's fine, I just want to see that you've at least read them. We can go from there.

**Comment by elalish** [comment]
I see no replies to any of my comment threads. Is there a chance they're pending and not submitted?

**Comment by clochardM33** [comment]
Ok. That was slightly embarrassing. Didn't know that comment visibility (pending/submitted) was linked into the file review actions. I've cost us a month, oops.

File: packages/model-viewer/src/three-components/SmoothControls.ts (hunk 1 above)
Comment (elalish) [reply packages/model-viewer/src/three-components/SmoothControls.ts]: Thank you for the explanation! That makes sense, but I'd rather fix the bug regarding incorrect zoom-out. You're absolutely right that it should be returning you to your initial view (with the exception of leaving theta and phi at their current values). It would probably be a good idea to add an example with this extreme FoV so we can be sure it's fixed and doesn't regress. It's also possible a small configuration change will help, if I can see what you're doing.

Yes, if you want selection, then single-click is nice for that, but we don't do selection in MV. Double-click we've found to be just too hard to discover, especially when single click does nothing.

File: packages/model-viewer/src/three-components/SmoothControls.ts (hunk 4 above)
Comment (elalish) [reply packages/model-viewer/src/three-components/SmoothControls.ts:998]: Looks like this? https://github.com/desktop/desktop/issues/557 The end of the thread has some things to try.

**Comment by elalish** [comment]
Okay, reading through all of this again, here's what I'd like to do. I want to see between the two of us, how much of what you want can be accomplished externally to MV. Thank you for giving it a shot and reporting your findings, but would you mind pasting that into a Glitch? Then we can work on it together and see the problems live. I have some ideas to work around the problems you found, and if we get it working, I think it could also make a nice addition to our examples. 

It's entirely possible we do in fact need MV changes, maybe bug fixes, maybe behavior or API changes, but I want to keep it minimal by checking against an example. Does that sound reasonable?

File: packages/model-viewer/src/three-components/SmoothControls.ts [hunk 7]
```diff
@@ -839,33 +864,168 @@ export class SmoothControls extends EventDispatcher {
...
+      opacity = 0;
+      // console.log('Released shift');
+    } else if (!event.shiftKey) {
+      opacity = 0;
+      // console.log('Released something else');
+    }
+    (this.scene.element as any)[$panElement].style.opacity = opacity;
+  }
+
+  private orbitZoomKeyCodeHandler(event: KeyboardEvent, ZoomInc: number, OrbitInc: number, isUserChange: boolean) {
```
[the remark this answers is not in the document]
Comment (elalish) [reply packages/model-viewer/src/three-components/SmoothControls.ts:935]: Thanks; would you mind writing out just the behaviors you desire for each user input? Then I at least know the UX you're going for and we can figure out the easiest way to accomplish it.

**Comment by clochardM33** [comment]
> 

That is all perfectly reasonable. I've got about a week before a couple of weeks of holidays catch up with the calendar.
So step 1 is get something on Glitch, step 2 is work through your other comments.
I'm presuming that on Glitch you are looking for things in pairs; The same model in m-v v12 & and then in m-v modified so you can compare and contrast behaviour?

**Comment by clochardM33** [comment]
Starter for 10

Using Release m-v
https://glitch.com/~iodized-delicious-night

Using Modified
https://glitch.com/~tulip-stream-grasshopper

File: packages/model-viewer/src/three-components/SmoothControls.ts (hunk 4 above)
Comment (clochardM33) [reply packages/model-viewer/src/three-components/SmoothControls.ts:998]: This has got to be something to do with it. Investigation under way.

File: packages/model-viewer/src/three-components/SmoothControls.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (clochardM33) [reply packages/model-viewer/src/three-components/SmoothControls.ts:935]: In words...
As the user I want to be able to use the keyboard to orbit, pan and zoom.
Because: I can already do all these things with touch and mouse, but only pan with keys and for some, but not all models, being able to step a change, rather than drag a change, would be the preferred way of nudging or changing the view.

As a developer I want to be able to offer the keyboard interaction above and in addition offer two "steps sizes" as watching users interacting with the initial stepping controls reveals that there is no right size step and a single size is either too coarse or too fine. 

At the same time the idea of coarse/fine step sizes nicely applies to the mouse controls (in the sense that it's not weird and once seen gets an oh yeah, got it response).

**Comment by clochardM33** [comment]
There are a couple of replies against the files as requested.

File: packages/model-viewer/package.json [hunk 8]
```diff
@@ -51,7 +51,8 @@
     "watch:tsc": "tsc -w --incremental",
     "watch:rollup": "rollup -c -w --environment NODE_ENV:production",
     "watch:rollup:dev": "rollup -c -w --environment NODE_ENV:development",
-    "watch:test": "karma start"
+    "watch:test": "karma start",
+    "build:dev:serve": "npm run build:dev && npm run serve"
```
[the remark this answers is not in the document]
Comment (clochardM33) [reply packages/model-viewer/package.json:55]: Where do I "You need to leave a comment indicating the requested changes."???

File: packages/model-viewer/src/features/controls.ts (hunk 5 above)
Comment (clochardM33) [reply packages/model-viewer/src/features/controls.ts:379]: Where do I "You need to leave a comment indicating the requested changes."???

File: packages/model-viewer/src/three-components/SmoothControls.ts (hunk 4 above)
Comment (clochardM33) [reply packages/model-viewer/src/three-components/SmoothControls.ts:998]: This has got to be something to do with it. Investigation under way.

File: packages/model-viewer/src/three-components/SmoothControls.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (clochardM33) [reply packages/model-viewer/src/three-components/SmoothControls.ts:935]: In words...
As the user I want to be able to use the keyboard to orbit, pan and zoom.
Because: I can already do all these things with touch and mouse, but only pan with keys and for some, but not all models, being able to step a change, rather than drag a change, would be the preferred way of nudging or changing the view.

As a developer I want to be able to offer the keyboard interaction above and in addition offer two "steps sizes" as watching users interacting with the initial stepping controls reveals that there is no right size step and a single size is either too coarse or too fine.

At the same time the idea of coarse/fine step sizes nicely applies to the mouse controls (in the sense that it's not weird and once seen gets an oh yeah, got it response).

**Comment by elalish** [comment]
Okay, first fix regarding strange zoom-out behavior: Why not just set the max radius and max FoV the same as the initial radius and FoV? I can't see what you were trying to accomplish with 5m and 90deg.

**Comment by elalish** [comment]
Second, this glitch is really complicated; would you mind removing everything that's not the stuff we're talking about in this issue?

File: packages/model-viewer/src/three-components/SmoothControls.ts [hunk 9]
```diff
@@ -839,33 +878,121 @@ export class SmoothControls extends EventDispatcher {
...
+  }
+
+ /**
+  * Handles the Pan key presses
+  * @param event The keyboard event for the .key value
+  * @param dxy The magnitude of the move
+  * @param isUserChange Is this a user initiated event
+  * @returns boolean to indicate if the key event has been handled
+  */
+    private panKeyCodeHandler(event: KeyboardEvent, dxy: number, isUserChange: boolean) {
```
Comment (elalish) [inline packages/model-viewer/src/three-components/SmoothControls.ts:973]: The problem with reviewing this PR is it's too big and doing too many things. Let's split it up and work up slowly. Start with a PR that just adds keyboard pan functionality. As small a change as possible; no API change, no multiple rates. 

Then, you can make a separate one to make the keyboard sensitivity based on our existing sensitivity parameter. There we can experiment with the out-of-sync issues you were talking about and see if we can fix them. After that, we can look into the rest in separate PRs.

Does that sound good? Also a new PR is probably the best way to get rid of the weird diff stuff in this one.

**Review by elalish — CHANGES_REQUESTED** [review]

File: packages/model-viewer/src/three-components/SmoothControls.ts (hunk 9 above)
Comment (clochardM33) [reply packages/model-viewer/src/three-components/SmoothControls.ts:973]: Yes ok. 
Now to confirm, before clicking on the wrong thing yet again, I should close this pull request and start a new one?

(I've worked out that you can only have one PR open at anyone time)

**Comment by clochardM33** [comment]
> Okay, first fix regarding strange zoom-out behavior: Why not just set the max radius and max FoV the same as the initial radius and FoV? I can't see what you were trying to accomplish with 5m and 90deg.

Because it isn't a truism that the the initial view is the view bounds that I want.
The max-es are set to be different precisely so that I can get out beyond the initial view. As an example some of the models are of bits of kit that are ~1.5m tall, 2m long, 1m wide. The initial view is a reasonably tight 3/4 shot, but to view side on or to make a nice orbit you need to be able to pull back further than this initial view.

**Comment by clochardM33** [comment]
> 



> Second, this glitch is really complicated; would you mind removing everything that's not the stuff we're talking about in this issue?

Big lumps removed, but can take out more if you need me to. It's left with just the side panel and 4 buttons that jump to specific views. These are required (in some shape or other) to force the situation where you need to click back onto the model and hence the view-change part.
Keycontrols are independent of this button panel.

File: packages/model-viewer/src/three-components/SmoothControls.ts (hunk 9 above)
Comment (elalish) [reply packages/model-viewer/src/three-components/SmoothControls.ts:973]: Yes, let's go ahead and close this PR and open a new one (or possibly several). You are definitely allowed to open as many PRs as you like simultaneously. In practice, that's only a good idea if they are independent (no major merge conflicts between them).

**Comment by elalish** [comment]
> Because it isn't a truism that the the initial view is the view bounds that I want. The max-es are set to be different precisely so that I can get out beyond the initial view. As an example some of the models are of bits of kit that are ~1.5m tall, 2m long, 1m wide. The initial view is a reasonably tight 3/4 shot, but to view side on or to make a nice orbit you need to be able to pull back further than this initial view.

Okay, that makes sense. I think because the example given was extreme I was confused. Let's call that a separate bug, which you can file an issue on ideally. Let's fix that on it's own so that click returns to the start instead of the max.

**Comment by clochardM33** [comment]
Hi,
Sorry for the delay. Stuff going on then lack of WiFi at the holiday destination.
I have opened Pan keys 1 #3703 as a new PR to replace this one.
Started off small as instructed. Please advise in the new PR.
Thanks.

**Comment by elalish** [comment]
Thanks!
